File: lohas-papers-backend/app/services/paper_searcher.py
import asyncio
import logging

from app.external import pubmed, semantic_scholar
from app.models.schemas import QueryTransformResult, UnifiedPaper
from app.utils.deduplication import deduplicate_papers

logger = logging.getLogger(__name__)


async def search_all_sources(
    transform_result: QueryTransformResult,
    *,
    year_from: int | None = None,
    year_to: int | None = None,
    limit_per_query: int = 20,
) -> list[UnifiedPaper]:
    """Search both Semantic Scholar and PubMed in parallel for all academic queries,
    then merge and deduplicate the results."""

    queries = transform_result.academic_queries

    # Search queries sequentially to avoid rate limits (429),
    # but search both sources in parallel for each query.
    all_papers: list[UnifiedPaper] = []
    for i, query in enumerate(queries):
        if i > 0:
            await asyncio.sleep(1.0)  # Delay between queries to avoid 429

        tasks = [
            semantic_scholar.search_papers(
                query, limit=limit_per_query, year_from=year_from, year_to=year_to
            ),
            pubmed.search_papers(
                query, limit=limit_per_query, year_from=year_from, year_to=year_to
            ),
        ]
        results = await asyncio.gather(*tasks, return_exceptions=True)

        for j, result in enumerate(results):
            if isinstance(result, Exception):
                logger.warning("Search query %d source %d failed: %s", i, j, result)
                continue
            all_papers.extend(result)

    logger.info("Total papers before dedup: %d", len(all_papers))

    # Deduplicate
    unique_papers = deduplicate_papers(all_papers)
    logger.info("Total papers after dedup: %d", len(unique_papers))

    return unique_papers

File: lohas-papers-web/src/lib/backend/paper-searcher.ts
import * as pubmed from "./pubmed";
import * as semanticScholar from "./semantic-scholar";
import { deduplicatePapers } from "./deduplication";
import type { QueryTransformResult, UnifiedPaper } from "./types";

export async function searchAllSources(
  transformResult: QueryTransformResult,
  options?: {
    yearFrom?: number | null;
    yearTo?: number | null;
    limitPerQuery?: number;
  },
): Promise<UnifiedPaper[]> {
  const { yearFrom, yearTo, limitPerQuery = 20 } = options ?? {};
  const queries = getUniqueQueries(transformResult.academic_queries);
  const allPapers: UnifiedPaper[] = [];

  for (const query of queries) {
    try {
      const papers = await pubmed.searchPapers(query, {
        limit: limitPerQuery,
        yearFrom,
        yearTo,
      });
      allPapers.push(...papers);
    } catch (err) {
      console.warn(`PubMed search failed: ${query}`, err);
    }
  }

  const hasSemanticScholarKey = Boolean(process.env.SEMANTIC_SCHOLAR_API_KEY);
  if (hasSemanticScholarKey) {
    for (const query of queries) {
      try {
        const papers = await semanticScholar.searchPapers(query, {
          limit: limitPerQuery,
          yearFrom,
          yearTo,
        });
        allPapers.push(...papers);
      } catch (err) {
        console.warn(`Semantic Scholar search failed: ${query}`, err);
      }
    }
  } else {
    console.info(
      "Semantic Scholar search skipped: SEMANTIC_SCHOLAR_API_KEY is not configured",
    );
  }

  if (allPapers.length === 0) {
    const fallbackQuery = transformResult.original_query || queries[0];
    if (fallbackQuery && !queries.includes(fallbackQuery)) {
      try {
        const papers = await pubmed.searchPapers(fallbackQuery, {
          limit: limitPerQuery,
          yearFrom,
          yearTo,
        });
        allPapers.push(...papers);
      } catch (err) {
        console.warn(`PubMed fallback search failed: ${fallbackQuery}`, err);
      }
    }
  }

  console.info(`Total papers before dedup: ${allPapers.length}`);

  // Deduplicate
  const uniquePapers = deduplicatePapers(allPapers);
  console.info(`Total papers after dedup: ${uniquePapers.length}`);

  return uniquePapers;
}

function getUniqueQueries(queries: string[]): string[] {
  return Array.from(
    new Set(
      queries
        .map((query) => query.trim())
        .filter((query) => query.length > 0),
    ),
  );
}
